crates/biome_js_analyze/src/semantic_analyzers/nursery/use_for_of.rs

```diff
@@ -0,0 +1,384 @@
+use biome_analyze::{context::RuleContext, declare_rule, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_js_semantic::ReferencesExtensions;
+use biome_js_syntax::{
+    AnyJsExpression, AnyJsForInitializer, JsAssignmentExpression, JsAssignmentOperator,
+    JsBinaryExpression, JsBinaryOperator, JsForStatement, JsIdentifierBinding,
+    JsPostUpdateExpression, JsPostUpdateOperator, JsPreUpdateExpression, JsPreUpdateOperator,
+    JsSyntaxKind, JsSyntaxToken, JsUnaryOperator, JsVariableDeclarator,
+};
+use biome_rowan::{declare_node_union, AstNode, AstSeparatedList};
+
+use crate::{semantic_services::Semantic, utils::is_node_equal};
+
+declare_rule! {
+    /// This rule recommends a for-of loop when the loop index is only used to read from an array that is being iterated.
+    ///
+    ///
+    /// Source: https://typescript-eslint.io/rules/prefer-for-of/
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// for (let i = 0; i < array.length; i++) {
+    ///   console.log(array[i]);
+    /// }
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// for (let i = 0; i < array.length; i++) {
+    ///    console.log(i, array[i]);
+    ///  }
+    /// ```
+    ///
+    /// ```js
+    /// for (let i = 0, j = 0; i < array.length; i++) {
+    ///    console.log(i, array[i]);
+    ///  }
+    /// ```
+    ///
+    /// ```js
+    /// for (let i = 1; i < array.length; i++) {
+    ///    console.log(i, array[i]);
+    ///  }
+    /// ```
```

Comment:
This is **very** similar to the first example, we can remove it, unless we plan to show a different usage of `i`

Reply:
It's initialized with 1 and not 0.
I can remove it if I don't think it brings value to the example